### PR TITLE
[Environment Variable][1/N] Use thread-safe env variable API in c10

### PR DESCRIPTION
**Pull request by cyyever**
This PR is the beginning of attempts to wrap thread-unsafe getenv and set_env functions inside a RW mutex.

### CODE REVIEW

File: c10/util/env.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4996)
+#endif
+  // NOLINTNEXTLINE(concurrency-mt-unsafe)
+  auto envar = std::getenv(name);
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+  return envar != nullptr;
```
Comment (albanD) [inline c10/util/env.cpp:22]: envar.has_value()

File: c10/util/env.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#include <cstdlib>
+#include <shared_mutex>
+
+namespace c10::utils {
+
+static std::shared_mutex env_mutex;
+
+// Checks an environment variable is set.
+bool has_env(const char* name) {
+  std::shared_lock lk(env_mutex);
```
Comment (albanD) [inline c10/util/env.cpp:12]: I'm confused why this is a shared_lock and not a unique_lock for all functions?

File: c10/util/env.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4996)
+#endif
+  // NOLINTNEXTLINE(concurrency-mt-unsafe)
+  auto envar = std::getenv(name);
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+  if (envar) {
```
Comment (albanD) [inline c10/util/env.cpp:37]: != nullptr ?

File: c10/util/env.h [hunk 4]
```diff
@@ -1,41 +1,23 @@
...
 #include <optional>
+#include <string>
 
 namespace c10::utils {
+
+// Checks an environment variable is set.
+C10_API bool has_env(const char* name);
+
+// Reads an environment variable and returns the content if it is set.
+C10_API std::optional<std::string> get_env(const char* name);
```
Comment (albanD) [inline c10/util/env.h:13]: Should we document that we discourage use of this method for boolean checks in favor of check_env below to improve consistency of handling?

File: c10/cuda/CUDADeviceAssertionHost.cpp [hunk 5]
```diff
@@ -81,8 +82,8 @@ bool dsa_check_if_all_devices_support_managed_memory() {
 }
 
 bool env_flag_set(const char* env_var_name) {
-  const char* const env_string = std::getenv(env_var_name);
-  return (env_string == nullptr) ? false : std::strcmp(env_string, "0");
+  const auto env_string = c10::utils::get_env(env_var_name);
+  return !env_string ? false : env_string.value() == "0";
```
Comment (albanD) [inline c10/cuda/CUDADeviceAssertionHost.cpp:86]: Should we use check_env?

**Review by albanD — COMMENTED** [review]
Very cool!

**Review by malfet — CHANGES_REQUESTED** [review]
Can you please write a better description at what you are trying to achieve here?
I haven't read the standard, but being familiar with the implementation details, I don't see why getenv on its own is not thread safe.

It is probably unsafe when there are a concurrent `putenv` calls, but you are not adding that one to your env utils. 


[Edit] To quote https://en.cppreference.com/w/cpp/utility/program/getenv looks like it is part of the standard:
> This function is thread-safe (calling it from multiple threads does not introduce a data race) as long as no other function modifies the host environment. In particular, the POSIX functions [setenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/setenv.html), [unsetenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/unsetenv.html), and [putenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/putenv.html) would introduce a data race if called without synchronization.


[Edit2] Is this about suppressing some clang-tidy warnings or solving a real problem? If later, then solution is to provide thread-safe `extern "C"` implementation of those methods in the code, that would take over libc ones (all libc functions are weak symbols by design so that one can app can choose to override them) and provide thread safety from getenv calls fro say oneDNN, libcuda etc

**Comment by albanD** [comment]
I would expect that a user setting something on `os.env` in python would trigger such a race? That was my assumption here.

**Comment by malfet** [comment]
> I would expect that a user setting something on `os.env` in python would trigger such a race? That was my assumption here.

Yes, but the current implementation would do nothing to avoid it, as `putenv` call from CPython would not be serialized against `getenv` calls made from libtorch

File: c10/util/env.cpp (hunk 2 above)
Comment (cyyever) [reply c10/util/env.cpp:12]: Yes, I haven't added a putenv/setenv yet.

File: c10/util/env.cpp (hunk 1 above)
Comment (cyyever) [reply c10/util/env.cpp:22]: Here envar is char *.

File: c10/util/env.h (hunk 4 above)
Comment (cyyever) [reply c10/util/env.h:13]: Sometimes we want the specific contents of env variables for further string comparison, so they are used for different purposes.

File: c10/cuda/CUDADeviceAssertionHost.cpp (hunk 5 above)
Comment (cyyever) [reply c10/cuda/CUDADeviceAssertionHost.cpp:86]: Fixed

**Comment by cyyever** [comment]
@malfet I plan to add wrapper putenv/setenv in the shared lock in the write mode, so that 

> I would expect that a user setting something on `os.env` in python would trigger such a race? That was my assumption here.

A specific reproducing example may be hard to create. Also note that Torch may be used in a pure multi-threading C++ environment.

**Comment by cyyever** [comment]
@malfet @albanD set_env was added.

**Comment by cyyever** [comment]
> Can you please write a better description at what you are trying to achieve here? I haven't read the standard, but being familiar with the implementation details, I don't see why getenv on its own is not thread safe.
> 
> It is probably unsafe when there are a concurrent `putenv` calls, but you are not adding that one to your env utils.
> 
> [Edit] To quote https://en.cppreference.com/w/cpp/utility/program/getenv looks like it is part of the standard:
> 
> > This function is thread-safe (calling it from multiple threads does not introduce a data race) as long as no other function modifies the host environment. In particular, the POSIX functions [setenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/setenv.html), [unsetenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/unsetenv.html), and [putenv()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/putenv.html) would introduce a data race if called without synchronization.
> 
> [Edit2] Is this about suppressing some clang-tidy warnings or solving a real problem? If later, then solution is to provide thread-safe `extern "C"` implementation of those methods in the code, that would take over libc ones (all libc functions are weak symbols by design so that one can app can choose to override them) and provide thread safety from getenv calls fro say oneDNN, libcuda etc

The purpose is to suppress clang-tidy warnings and solve possible future problems by providing thread-safe getenv and setenv. Overwriting libc implementations may not be a good idea because we have to support MSVC and other no GNU platforms so that I prefer to use explicit c10 API.

File: c10/util/env.h (hunk 4 above)
Comment (albanD) [reply c10/util/env.h:19]: Ho yes, that why I'm saying that this one should be used only for non-boolean env.

Comment (cyyever) [reply c10/util/env.h:19]: I see what you mean, let's add a comment.

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by PaliC** [comment]
@pytorchbot revert -m "your using TORCH_INTERNAL_ASSERT incorrectly" -c "ghfirst"

We're getting the following error from our compiler
```
xplat/caffe2/c10/util/env.cpp:37:9: error: implicit conversion turns string literal into bool: 'const char[32]' to 'bool' [-Werror,-Wstring-conversion]
        "setenv failed for environment \"", name, "\", the error is: ", err);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
xplat/caffe2/c10/util/Exception.h:406:31: note: expanded from macro 'TORCH_INTERNAL_ASSERT'
  if (C10_UNLIKELY_OR_CONST(!(cond))) 
  ```
  
  fixing the assert to start with a bool should fix things.

**Comment by cyyever** [comment]
@PaliC Thank you for pointing it out, fixed.

**Comment by malfet** [comment]
Though I might be wrong, just a lots of unrelated failures, sorry about the revert

**Review by malfet — APPROVED** [review]

**Comment by cyyever** [comment]
@jeanschmidt Some hints about the internal failures?

**Comment by albanD** [comment]
@zdevito is looking into similar issues. I will let him comment here on what is the easiest way forward.

**Review by eqy — APPROVED** [review]

**Comment by cyyever** [comment]
@zdevito  Can you help?

**Comment by eqy** [comment]
commenting that this is a real issue we've seen in practice https://github.com/pytorch/kineto/pull/984

but problematic interactions might be more likely between PyTorch and third party libraries as above

**Comment by cyyever** [comment]
Rewrite code to avoid changing existing API.

**Pull request closed without merging**